pkg/controllers/managementuser/clusterauthtoken/token.go

```diff
@@ -34,14 +36,116 @@ type tokenHandler struct {
 	userAttributeLister        managementv3.UserAttributeLister
 	clusterSecret              corev1.SecretInterface
 	clusterSecretLister        corev1.SecretLister
 }
 
+// extCreate is called when a given ext token is created, and is responsible for
+// updating/creating the ClusterAuthToken in the downstream cluster.
+func (h *tokenHandler) extCreate(token *extv1.Token) (*extv1.Token, error) {
+	logrus.Debugf("[%s] ext CREATE FOR %q INTO %q", clusterAuthTokenController, token.Name, token.Spec.ClusterName)
+
+	_, err := h.clusterAuthTokenLister.Get(h.namespace, token.Name)
+	if !errors.IsNotFound(err) {
+		return h.ExtUpdated(token)
+	}
+
+	// we can sync tokens which are hashed by copying the hash downstream
+	// token is hashed, we can safely attempt to sync downstream
+	hashVersion, err := hashers.GetHashVersion(token.Status.Hash)
+	if err != nil {
+		// the token hash is unlikely to change, re-enqueing would just produce a flood of errors
+		logrus.Errorf("unable to determine hash version of token [%s], will not sync token: %s", token.Name, err.Error())
+		return token, generic.ErrSkip
+	}
+	// we only sync tokens downstream that were created with SHA3
+	if hashVersion == hashers.SHA3Version {
+		return nil, h.createClusterAuthToken(token, token.Status.Hash)
+	}
+
+	// token is hashed, but we can't sync it since we don't have the raw value
+	logrus.Warnf("token [%s] will not be synced or useable for ACE because it uses an older hash version, generate a new token to use ACE", token.Name)
+	// don't re-enqueue, we can't sync this token
+	return nil, generic.ErrSkip
+}
+
+// ExtUpdated is called when a given ext token is modified, and is responsible
+// for updating/creating the ClusterAuthToken in a downstream cluster.
+func (h *tokenHandler) ExtUpdated(token *extv1.Token) (*extv1.Token, error) {
+	logrus.Debugf("[%s] ext UPDATE FOR %q INTO %q", clusterAuthTokenController, token.Name, token.Spec.ClusterName)
+
+	clusterAuthToken, err := h.clusterAuthTokenLister.Get(h.namespace, token.Name)
+	if errors.IsNotFound(err) {
+		return h.extCreate(token)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	err = h.updateClusterUserAttribute(token.GetUserID())
+	if err != nil {
+		return nil, err
+	}
+
+	tokenEnabled := token.GetIsEnabled()
+	current := tokenAttributeCompare{
+		enabled:   tokenEnabled,
+		expiresAt: token.Status.ExpiresAt,
+		username:  token.Spec.UserID,
+	}
+	old := tokenAttributeCompare{
+		enabled:   clusterAuthToken.Enabled,
+		expiresAt: clusterAuthToken.ExpiresAt,
+		username:  clusterAuthToken.UserName,
+	}
+
+	// note: ext tokens are always hashed (contrary to v3 Tokens)
+	hashVersion, err := hashers.GetHashVersion(token.Status.Hash)
+	if err != nil {
+		logrus.Errorf("unable to determine hash version of token [%s], will not sync token: %s", token.Name, err.Error())
+		return token, generic.ErrSkip
+	}
+	// we only sync tokens downstream that were created with SHA3
+	if hashVersion == hashers.SHA3Version {
```

Comment:
There are no tests for this logic?

Reply:
Same logic as for v3 tokens, lines 287 to 297.
That said, checking shows that the v3 logic is tested, and that I have no tests for the ext logic at all.
Will be fixed.

Reply:
Done, commit 7faac8db94ae39033191821e24752c5c2049d9aa
Caught missing functionality.
`ExtUpdated` was not updated to deal with the new side-secret holding the token hash.
Fixed in commit 34e6ad12107b0c951dcb277f87d5fe5e4c3fb535